Initial merge bot implementation.

eduard-bagdasaryan opened the pull request.

eduard-bagdasaryan committed:
Fixed eslint errors except 'defined but never used'.

The 'no-unused-vars' rule currently set to 'error'.  There are some
callbacks passed to another libraries, having such variables. I don't
think we should change passed callbacks signature, because these
variables (e.g., error messages) are unused for now, but may be used in
future.

eduard-bagdasaryan committed:
Polishing: substituted with arrow function expressions.

eduard-bagdasaryan committed:
Reworked eggtimer bot to conform to requirements.

AFAICT, the requirements are:

1. Do not cache any information received from Github, except data used
   for processing a single PR.

2. Do not use received events data, since it is incomplete and requires
   caching for its processing.

3. Use events only as triggers to start polling all available open PRs.

4. Configure Github repo/owner in order to do PRs polling on bot
   startup.

This revision does not change 'merge' functionality yet (including
auto_branch operation).

eduard-bagdasaryan committed:
Initial 'PR merging via auto_branch' implementation.

Not fully tested yet.

eduard-bagdasaryan committed:
Fixed some bugs after testing.

eduard-bagdasaryan committed:
Re-worked merging algorithm, using Github 'commit' API.

The previous implementation, which was based on Github 'merge' API,
has a flaw: after merge into auto_branch is done, there is no way
to undo it. It makes no problem if auto_branch tests are successful,
but if they are not, we end up with closed PR and inability to
finish our merge into master.

There is no ideal solution for this problem, so our trade of is
the following: do not use 'merge' button API and do manual merge
instead. The disadvantage is after merging the PR gets only
'closed' status (instead of 'merged' and 'closed'), confusing the
reader that the PR became abandoned for some reason. To eliminate
this problem we decided to mark the PR with a label(not done yet).

This solution makes use of Github 'merge' revision, stored as
a reference under 'refs/pull/:pr/merge' reference path, which
is created by Github for each PR. With Github 'commit' API,
we create a new commit, pointing to this 'merge' revision and reset
auto_branch into it. If checks are successful, we merge master
with FF mode into 'auto', closing the PR.

eduard-bagdasaryan committed:
Mark closed PRs with "S-merged" label.

eduard-bagdasaryan committed:
Mark merging PR which auto checks were failed.

If PR's auto branch checks failed during merge, the PR is has still
'mergeable' status and is scheduled again and again.  I mark such PRs
with a label 'S-merge'failed' in this case.  When the underlying
problems are fixed, this label can be removed manually, thus activating
PR merging again.

Also: unified PR numbering in debug messages.

eduard-bagdasaryan committed:
Do not use Github 'merge' revision, create it by hand.

Using 'merge' revision, supplied by Github does not work if master
changes: we don't want to bother user pressing "Update branch" on the
PR. So I have to create that 'merge' revision separately, using master's
HEAD for that.

eduard-bagdasaryan committed:
Revert "Do not use Github 'merge' revision, create it by hand."

This reverts commit c2a42388bebd45e8867548c495c2516a3a28dc4d.
That change was based on the incorrect assumption that PR 'merge'
revision is not updated if master changes.

eduard-bagdasaryan committed:
Reworked non-auto_branch code with promises.

With this refactoring, the code looks simpler and more reliable.

eduard-bagdasaryan committed:
Restore merging context on startup.

Now, when bot starts, it checks whether there is a still merging
auto_branch revision. If there is one, it completes merging process.
This is needed primary for a situation when the bot was terminated while
it was waiting for auto_branch checks completing (for a PR had been
merged into auto_branch).

eduard-bagdasaryan committed:
Restore merging context on startup (pt. 2).

Added functionality for commenting being-in-merge commit and dropping
this comment after merge into master is completed.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Bug fixing.

eduard-bagdasaryan committed:
Properly handle previous unsuccessful merge attempts.

Before this change, such PRs (which are marked with "S-merge-failed"
label) were ignored, until somebody drops this label manually.
Now, a PR with "S-merge-failed" is considered as a merge candidate if
its 'merge' revision has been changed since last unsuccessful merge
attempt (in hope the underlying problem has been fixed by these
changes).  Github automatically updates 'merge' revision if
changes either PR branch or its base branch(usually master).

eduard-bagdasaryan committed:
Fixed several problems after testing.

* PR 'merge' commit is not updated automatically when
  master changes. To make this happen, we should get the PR via API.

* Distinguish 'ref' parameter in
  createReference/updateReference/deleteReference API calls.

* Do not reset 'currentContext' before its 'signaled' flag is processed.

* Properly manage with 'merging-PR' tag: update if necessary and delete
  after merge completion.

* Few other bugs.

eduard-bagdasaryan committed:
Updated config template with actual parameters.

eduard-bagdasaryan committed:
More tests performed and fixed found problems.

* When getting PR information about 'mergeable' flag,
  this information may be still unavailable, meaning that Github is
  currently calculating this 'mergeability'. In this case, we should
  repeat request, until Github returns non-null 'mergeable'.

* Remove "S-merge-failed" label after problems are fixed and the PR
  is merged.

eduard-bagdasaryan committed:
Refactoring: use 'merging tags' instead of 'merging commit comments'.

Both mechanisms were excessively used for marking such 'being-in-merge'
commits. This marking is needed to be able to restore merging context if
something goes wrong during merge, for example, bot terminates or PR
checks fail. This change eliminates commit commenting, because tagging
is enough.

eduard-bagdasaryan committed:
Refactored and improved logging.

* Unifying messages, generated after promisificated methods completion
  (i.e., 'resolve' or 'reject' detailed information).

* Several new helper methods and objects used to collect and output
  exception messages.

* Improved messages wording.

eduard-bagdasaryan committed:
Refactoring: grouped related methods, no functionality changes.

eduard-bagdasaryan committed:
Addressed several review comments.

eduard-bagdasaryan committed:
Big refactoring, according to the review and discussion.

The major work completed, few tests passed.

There are still few not_addressed/not_completed questions,
such as configuration improvement and better error handling.

eduard-bagdasaryan committed:
Big refactoring, according to the review and discussion, pt. 2.

This change addresses questions, not completed in f5d9f72, such
as creating class wrappers for 'Config' or 'Context' entities,
simplifying logging, and fixing some found problems.

eduard-bagdasaryan committed:
Properly handle possible getPRList() rejections.

eduard-bagdasaryan committed:
Code improvements after review:

1. Get PR list in runStep(), instead of run(). runStep() algorithm is
   the following now: select PRs from the list one by one and process
   them, until either:

   * there is a PR we need to wait to,
   * we processed all PRs from the list,
   * an unexpected error occurred.

2. Protect ConfigOptions fields from typos: use 'public' methods
   instead.

3. Re-run in 10 minutes (default) in a case of unexpected error.

eduard-bagdasaryan committed:
Several improvements after discussion.

* Get 'required' status checks from Github (instead of configuring
  required status checks number).

* Get reviewers list from Github. For now, the bot keeps using the
  existing 'reviewers_number' parameter, but analyses the reviewers
  having 'push' right (these are considered as valid approvals).

* Use 'rerun' flag in 'run' instead of 'runStep' for simplification.

* A few other fixes, e.g., missed 'await' in cleanupOnError().

eduard-bagdasaryan committed:
Code refactoring.

* Made exception handling style more consistent:
  handle most of exceptions in high-level runStep(), instead of
  startMerging(), finishMerging(), etc. Made cleanupOnError()
  non-throw for this purpose.

* Distinguish FF-merge failure, checking for 422 HTTP error. Re-run
  merge, if such PR was the last one in the list.

* getPR() optimization: do not wait while Github calculates
  'mergeable' flag, if we do not need it(yet).

* Polishing: removed some stale code, renamed RejectMsg -> ErrorContext,
  etc.

eduard-bagdasaryan committed:
Fixed auto_branch parameter example.

eduard-bagdasaryan committed:
Refactored with classes.

The main purpose of this refactoring was to get rid of global variables,
such as 'PRList' and 'Context'. The code was split between two classes:
'MergeStep', which processes entire PR list from Github and
'MergeContext', which processes a single PR.

Also made some corrections into the merging algorithm:

* Check 'auto_branch' once before processing PR list, instead of doing
  this before processing each PR from the list.
* Do not reset auto_branch into master in case of error (the error
  may be caused by one PR, but auto_branch point to another).

eduard-bagdasaryan committed:
Use correct 'base' branch instead of hard-coded 'master'.

eduard-bagdasaryan committed:
Made vote analysis essentially conforming to the official 'MergeProcedure'.

* A 'negative' vote blocks the merge.
* Approved PRs are submitted after a period (10 days by default).
* If PR author is a 'core' developer, his vote also is considered.

A new 'RunScheduler' class contains old globals and new rerun-planning
code.

eduard-bagdasaryan committed:
High-level exception handling improvement.

* Unified exception and error handling in Scheduler.run().

* In case of such high-level(unpredictable) exception, ignore other
  events during the period(10 min by default). This should avoid
  wasting resources or filling logs with much noise.

eduard-bagdasaryan committed:
Log messages with 'bunyan'.

eduard-bagdasaryan committed:
Two-step vote analysis.

* After PR creation, there is a two-days(default) period
  to wait for a core-developer review.

* If there is only one core-developer review, still wait for
  10 days approval period.

eduard-bagdasaryan committed:
Labeling improvements.

Made label manipulation more correct, each PR 'state' corresponds to
a label:
* 'merging'
* 'merge-ready'
* 'merged'
* 'merge failed'
* 'auto checks failed'

Several other fixes and polishing after testing.

eduard-bagdasaryan committed:
Adjusted README accordingly.

eduard-bagdasaryan committed:
Adjusted README markup

eduard-bagdasaryan committed:
README.md: whitespace fixes.

eduard-bagdasaryan committed:
Several fixes after testing.

* Properly honour 10 min. recover delay. The approach 'select other PR
  if this one has failed' had an unhandled case when all PRs have failed
  so that there are no more PRs to select.

* Avoid unnecessary cleanup procedure on error (adjusting labels,
  deleting tags) for 'skipped' PRs (PRs with autochecks failed status).
  Though useless setting up labels has a small performance overhead,
  deleting a tag (with autochecks failed status) is more expensive
  because it will cause another merge iteration with the predicted
  autochecks failed result.

* Refactored and fixed user voting functionality. The previous
  implementation had some bugs (due to the newly added
  'reject timeout' rule) and became too complex.

* Do not mix up 'promisificated methods' with other bot logic (as
  all other methods follow to). Extracted all such logic from
  'getReviews()' and 'getStatuses()' into a separate checkApproved()
  and checkStatuses(). Also reworked and simplified checkStatuses().

* Polished several other places and code comments.

eduard-bagdasaryan committed:
Simplified single-parameter 'arrow' functions.

eduard-bagdasaryan committed:
Updated 'reject period' definition.

eduard-bagdasaryan committed:
Getting statuses improvement.

The initial version of the bot used getStatuses() API call for getting
all PR statuses. There is a better "combined statuses" call for our
purposes, returning the most recent statuses.

eduard-bagdasaryan committed:
Clarified auto_branch use cases and other fixes.

1. Avoid useless auto_branch switches.
   Before this change, auto_branch reference was adjusted every time
   PR's tag was analyzed. This was not a problem initially, but after
   adding 'skip_merge' mode (a testing mode when the last fast-forward
   step is skipped) revealed a problem with multiple PRs:
   switching auto_branch between PRs caused endless Github 'push' events
   and PRs status updates. Switching auto_branch to 'master' (in case of
   an error) was another cause of such unwanted PUSH event.

2. Do not use auto_branch SHA for fast-forwarding, use merge tag SHA
   instead (due to (1)). With this change, auto_branch has two purposes:
   * Mark the PR's merge tag for signaling CI systems.
   * Find the being-in-merge PR (i.e., the merging tag auto_branch
     points to).

3. Simplified, eliminating 'autoStarted' and 'inProcess' flags.

4. Several other fixes and commenting.

eduard-bagdasaryan committed:
Dry-run polishing.

* Protect with assertions modifiers API calls, such as
  updateReference, deleteReference etc.

* Eliminated logging code duplication.

eduard-bagdasaryan committed:
Fixed a case when no required checks configured.

Also fixed a case when no statuses have been calculated yet
(status list is empty). Obviously, this is a 'pending' state.

eduard-bagdasaryan committed:
A typo in _checkStatuses() and several polishing fixes.

eduard-bagdasaryan committed:
Corrected definitions for several parameters.

eduard-bagdasaryan committed:
Minor adjustments after discussion.

* Start PR approval period from the PR creation date (instead
  of the first approval date).

* React to the Github 'push' event, since we need to rerun
  when base branch (master) changes.

rousskov committed:
Rewrote with a focus on users/admins. Needs more work.

eduard-bagdasaryan committed:
Renamed Github -> GitHub.

eduard-bagdasaryan committed:
Configurable Logger parameters, create Logger just after configuration.

eduard-bagdasaryan committed:
Configure HTTP server with 'host' parameter (INADDR_ANY, if empty).

eduard-bagdasaryan committed:
Changed config extension to '.json'.

eduard-bagdasaryan committed:
Configure 'approval' and 'reject' periods in hours (instead of days).

eduard-bagdasaryan committed:
Renaming 'auto' -> 'staging'.

eduard-bagdasaryan committed:
Eliminated 'MergingTagName' global.

eduard-bagdasaryan commented on README.md:
```diff
@@ -1,68 +1,256 @@
 
-Create a config.js, by cloning config-example.js:
+* The PR is open (TODO: Check that a PR closed right _after_ we got a
+  list of PRs will not be merged!)
```
This check exists in a low-level requestPR().

rousskov replied:
Agreed. Please move that check out of that low-level method into the method that tests preconditions. Low-level methods should not make assumptions about additional (i.e., not expressed in method signature and description) caller needs.

rousskov replied:
BTW, please rename getPR() and/or requestPR().

eduard-bagdasaryan replied:
Moved this check into higher level code.
BTW, how do you suggest to rename getPR/requestPR()?

rousskov replied:
> how do you suggest to rename getPR/requestPR()?

I only have a negative suggestion for now: Please do not use essentially the same name for two different functions. Needless to say, the names should reflect what the functions do. If you cannot find proper names, I will suggest something specific after the code changes are done (when the "what the functions do" part becomes stable).

eduard-bagdasaryan commented on README.md:
```diff
@@ -1,68 +1,256 @@
+* The PR is open (TODO: Check that a PR closed right _after_ we got a
+  list of PRs will not be merged!)
+* The PR has GitHub 'mergeable' status. TODO: Document how a GitHub user
+  can tell whether a PR has a GitHub 'mergeable' status.
```
I assume a user can understand this if there is no 'This branch has conflicts...' message. I can add a screenshot to illustrate, if needed.

eduard-bagdasaryan commented on README.md:
```diff
@@ -1,68 +1,256 @@
+  branch commit. The bot removes this label before it updates the PR
+  target branch. (TODO: Remove due to very short lifetime and being low
+  on information? All S-merged and S-merge-failed PRs are S-merge-ready
+  for a brief moment...)
```
This label primary purpose is to mark our 'dry run 2' mode (when all staging checks have passed but we skipped the last step). Without dry run, it will disappear very quickly, as you noted.

eduard-bagdasaryan replied:
I fixed this in a way we discussed: mark only when 'dry run 2' mode is on.

eduard-bagdasaryan commented on README.md:
```diff
@@ -1,68 +1,256 @@
+manual PR merging, of course.
+
+TODO: Should the bot recheck the conditions again after all staging
+branch tests are over?
```
I don't think it should. I master changes, the FF merge fails and we will notice that. If branch changes, we should not care about that: we merge one of its the previous revisions.

eduard-bagdasaryan replied:
Done at 3a063d4.

eduard-bagdasaryan committed:
Get a fresh PR list from Github at each step.

* Before this fix there were few cases, when the PR list
  could become a bit stale (we finished merging and used previously
  cached PR list).  Though such situations could hardly happen
  in reality (fast-forwarding operation is fast), this change makes the
  code more clear.

* Do not manipulate PR list when not necessary. This covers a situation
  when we have to resume a being-in-merge PR processing and should not
  get this list before we finish with it.

* Refactored code responsible for 'merge tag' analysis, distinguishing
  two different PR modes: before merge (a new precondition) and
  being-in-merge (checking for auto-tests results).

eduard-bagdasaryan committed:
Sort PR list by dates in ascending order.

GitHub returns PRs in descending order.

eduard-bagdasaryan committed:
Use a single(soonest) timer for PR planning

eduard-bagdasaryan committed:
Mark with 'S-merge-ready' label only in 'skip merge' debug mode.

eduard-bagdasaryan committed:
Properly check whether a PR was unexpectedly closed.

Though the current bot implementation does not deal with 'closed' PRs,
putting such check inside low-level getPR() violates a simple approach:
promisificated wraps do not check for non-error results, passing them to
callers.

eduard-bagdasaryan committed:
Validate merge conditions once more just before fast-forwarding.

Though all these checks have been validated at the beginning of PR merge
process, something may happen during this processing (e.g., somebody
requested more changes). We need to reject such merge attempts
(which would succeed otherwise).

For simplicity, The 'post-condition' check will only work for cases when
the 'merge revision' itself has not been changed (i.e., neither master,
no PR branch have been changed).

eduard-bagdasaryan committed:
Clarified scope separation between RunScheduler and Web server.

eduard-bagdasaryan committed:
Split the monolithic script into several modules.

Also fixed stack trace logging in ErrorContext.

eduard-bagdasaryan committed:
Simplified PR planning, using a single timer on each iteration.

PR Planning is used for ready-for-merge PRs, approved by a single core
developer. Squid voting rules prescribe to wait for a period (10 days
from creation date) in this case. As a consequence, we do not need to
analyze all such PRs, planning a single one, created first.

eduard-bagdasaryan committed:
Resolving modules cyclic dependencies.

* Refactored code, responsible for delays in order to eliminate
  RepoMerger <-> PrMerger dependency.

* Placed global 'Rerun' flag into a separate module.

Also extended 'post-condition' check (3a063d4) to cover PR branch
changes.

eduard-bagdasaryan committed:
Re-start HTTP server on unexpected error.

eduard-bagdasaryan commented on README.md:
```diff
@@ -1,68 +1,256 @@
+   message. TODO: What happens to long paragraph lines often present in
+   the PR description? We should either reject PRs that have long lines
+   or, if it can be done safely, automatically wrap them using some
+   formatting library.
```
It would be better(safer) to reject such PRs with long lines and PRs with GitHub markdown. I have read several guidelines about writing commit messages, all of them prescribe to keep to 72 characters per line.
Alternatively, we can try to automatically format these messages with some available tools: e.g., firstly [remove](https://github.com/stiang/remove-markdown) markdown and then [wrap](https://www.npmjs.com/package/word-wrap) the message. However, I have not tested these tools and don't know whether they are reliable.

rousskov replied:
For now, I suggest disqualifying PRs that use excessively long lines (>79 chars?) and allowing/using everything else "as is". We will probably need a new label to mark/identify affected PRs (in addition to the merge-failed label that will also be applied).

The following items become TODOs for future work, in rough priority order:

1. Auto-convert markdown to plain text during merge, before checking for long lines. This will take care of most long lines, verbatim block quotes, and enumerated lists.
1. Provide a way to disable this check via a PR comment command. Some long lines should be admitted.
1. Provide a way to auto-convert the PR description via a PR comment command, with the original text posted as a comment for backup purposes.
1. Prohibit undesirable markdown while still allowing simple things like `_foo_` and `*bar*`.

eduard-bagdasaryan committed:
Reject PRs with long lines in description

eduard-bagdasaryan committed:
Properly handle a situation when the bot was terminated during cleanup

When the PR staging commit was merged on it's base, there are still some
cleanup steps to be executed, such as PR labeling, closing and removing
tag. After this fix, the bot will try to complete these cleanup steps
even if it was killed during them(or an unexpected error occurred).

eduard-bagdasaryan committed:
Get rid of global Rerun flag

This flag was improperly used by low-level code in two contexts:

1. It was checked just before startMerging(). Actually, this was a minor
   optimization, because PR list is processed rather quickly (several
   seconds) and we probably should not re-run during this operation.

2. It was assigned in MergeContext when getting a fast forwarding error
   in order to re-run from scratch. This probably should be done in a
   higher level, without changing the logic itself.

Also keep dry_run mode readonly: do not label PRs having invalid
title/description.

eduard-bagdasaryan committed:
Properly handle unexpected terminations on resume merging

This is a second part of modifications, started in 348a9c0.  Before this
change, all unexpected errors during resuming PR merging caused a
general cleanup: removing the tag and adjusting labels (Merge-failed).
However such approach is not reliable and can leave the PR in an
inconsistent state.  For example, if an error occurred during
fast-forwarding, we end up with still open PR but actually merged PR.
If an accidental error occurs before fast-forwarding step(while
extracting information about 'current' PR), we can't remove the
tag before we are sure that the tag revision is not in master already
(in a way without bot participation). Also, it looks better to repeat
merge attempt (after 10 minutes) instead of immediately removing the tag
and discarding valuable staging test results(which could last for
hours).

This change addresses a TODO from 348a9c0: compare commits in order
to properly check whether a revision has been merged.

eduard-bagdasaryan committed:
Renamed MergeStep -> PrMerger

eduard-bagdasaryan committed:
Reject PRs with requested reviewers, if they are core developers

eduard-bagdasaryan committed:
Re-start HTTP server on an unexpected error(pt. 2)

This is an improvement over aae8c98: we should re-create the HTTP server
object for reliability sake (instead of just closing/starting the
already existing server object).

eduard-bagdasaryan committed:
Reworked getPR/requestPR naming duplication

eduard-bagdasaryan committed:
Pass bot configuration file name as a command line argument

eduard-bagdasaryan committed:
Renamed configuration parameters, such as voting_delay_min/max, etc.

eduard-bagdasaryan committed:
Checked/fixed README

eduard-bagdasaryan committed:
Removed debugging assert

eduard-bagdasaryan committed:
Simplified, removing MergeContext::ffMergeFailed flag

eduard-bagdasaryan committed:
A couple of fixes

* Missed 'await' at asynchronous _labelCheckMessage().

* Fixed logError() to properly handle unusual exceptions such as
  strings.

eduard-bagdasaryan committed:
Added screenshots to README

It is better to generate suitable images because those images from
GitHub help pages are large and have some excessive information.

eduard-bagdasaryan committed:
Fixed images placement in README

eduard-bagdasaryan committed:
Several minor fixes

* Adjusted 'host' parameter description in README.
* Adjusted config-example.json.
* Few other polishing fixes.

eduard-bagdasaryan committed:
RepoMerger: clarified PR planning and rerunning

eduard-bagdasaryan committed:
Polished PrMerger class

eduard-bagdasaryan committed:
Polished MergeContext class

eduard-bagdasaryan committed:
A couple of typos fixed

eduard-bagdasaryan committed:
Simplified PrMerger.current()

eduard-bagdasaryan committed:
More polishing for MergeContext methods

eduard-bagdasaryan committed:
Clarified MergeContext.delay()

eduard-bagdasaryan committed:
A typo in method name

eduard-bagdasaryan committed:
A couple of improvements

* Pass raw JSON configuration to the bunyan constructor
* README: fixed screenshot and phrasing

eduard-bagdasaryan committed:
README: indentation and punctuation

eduard-bagdasaryan committed:
README: a couple of fixes

eduard-bagdasaryan committed:
Getting protected check list with getBranch API

... instead of more specific getProtectedBranchRequiredStatusChecks().
For unknown reason the latter one did not work for squid repository
and anubis bot (though the bot was granted all required permissions).

eduard-bagdasaryan committed:
Handle a 404 Not Found response when getting required checks list

eduard-bagdasaryan committed:
Removed unnecessary _logError() method

eduard-bagdasaryan committed:
Removed unnecessary "_branch" prefix for auto branch

rousskov committed:
Revised README. Use label names with a better "structure".

Also added two new labels (to be implemented eventually).

Also spotted and documented to XXXs to be fixed.

eduard-bagdasaryan committed:
Do not add 'merge failed' label when staging checks failed

... but use a more specific 'staging failed' label, as documented in
README.
Do not set 'staging failed' label every time the failed tag was found:
this setting looks redundant.
Also distinguish different 'merging commit' and 'staging commit'
notions.

eduard-bagdasaryan committed:
Made logging more consistent in checkMergeConditions()

eduard-bagdasaryan commented on README.md:
```diff
@@ -93,75 +105,106 @@ internal GitHub server errors), the bot sleeps for 10 minutes
+  lying in documentation, we should either remove this label only when
+  (re)setting the first label (i.e., M-waiting-staging-checks) or remove
+  this label when setting any other label, without making an exception
+  for M-failed-staging-checks. The implementation plans for the
```
I would not do as you suggested, because:
* "removing the label only when resetting the first label" misses a case when we should remove it after PR is merged(and closed): we should not keep such excessive details after PR completion.
* "removing when setting any other" (including M-failed-staging-checks) also looks excessive because I do not see how the passed checks suddenly become 'failed'.

Instead, I would suggest fixing existing code and the documentation: IMO, we need to remove this label only in two cases:
* M-merged
* M-failed-other
because these should be the only possible states for PR after it's staging checks have succeeded.

rousskov replied:
> "removing the label only when resetting the first label" misses a case when we should remove it after PR is merged

I would expect the first label (i.e., `M-waiting-staging-checks`) to be (re)set when 

* the bot (re)starts waiting for those checks (because something has changed and new tests are needed) or 
* the bot decides to proceed beyond staging checks (because the `config::skip_merge` option was removed). Since the bot does not read/check `M-waiting-staging-checks`, it has to reset it when proceeding beyond the staging checks, right?

Both events must happen way before the PR is merged. Am I missing something?

> "removing when setting any other" also looks excessive 

It certainly is excessive! However, it is also very simple. If removing a non-existent label does not generate a user-visible notification on GitHub, then it is a viable option because of its simplicity.

> IMO, we need to remove this label only in two cases: `M-merged` and `M-failed-other`

... and `M-waiting-status-checks` and `M-failed-description` (at least).

If you want to minimize `M-passed-staging-checks` removals, then we should remove this label when it is set but is no longer valid. It is no longer valid when the _tested_ commit is no longer fresh/applicable. I do not recommend this path because it probably increases code complexity.

eduard-bagdasaryan commented on README.md:
```diff
@@ -93,75 +105,106 @@ internal GitHub server errors), the bot sleeps for 10 minutes
+* `M-failed-other`: A fatal PR-specific error other than the staging
+  branch test failure (the latter is marked with
+  `M-failed-staging-checks`). (XXX: finishProcessing() sets this label
+  even if the staging branch test fail). It is probably necessary to
```
I did not understand this XXX: finishProcessing() should be the only place where we know about failed staging checks. Probably another method was meant here, which was fixed at 19f4956.

rousskov replied:
Thank you for fixing this.

Rephrasing XXX FYI: finishProcessing() sets `M-failed-other` when the staging branch test fails. finishProcessing() should not do that.

eduard-bagdasaryan committed:
Few fixes in README

eduard-bagdasaryan committed:
Adjusted label names according to README

eduard-bagdasaryan committed:
Clarified M-passed-staging-checks description (removing XXX)

... and do not remove labels if not required(optimized excessive
removeLabel() API calls).

rousskov commented on Config.js:
```diff
@@ -0,0 +1,66 @@
+    waitingStagingChecksLabel() { return "M-waiting-staging-checks"; }
+    // Merge succeeded up to fast-forward step. For testing purpose.
+    passedStagingChecksLabel() { return "M-passed-staging-checks"; }
+    // PR message does not satisfy to requirements (e.g., lines
```
Please use this description instead: `future commit message violates requirements`

eduard-bagdasaryan replied:
Done.

rousskov commented on Config.js:
```diff
@@ -0,0 +1,66 @@
+    mergedLabel() { return "M-merged"; }
+    // merge started (tag and staging branch successfully adjusted)
+    waitingStagingChecksLabel() { return "M-waiting-staging-checks"; }
+    // Merge succeeded up to fast-forward step. For testing purpose.
```
Please use this description instead: `passed staging tests (in staged-run mode)`

eduard-bagdasaryan replied:
Done.

rousskov commented on config-example.json:
```diff
@@ -0,0 +1,28 @@
+    "repo": "squid",
+    "owner": "squid-cache",
+    "dry_run": false,
+    "skip_merge": false,
```
Please rename `skip_merge` (and equivalent names) to `staged-run` (and equivalent names). Please do not forget to update README accordingly.

eduard-bagdasaryan replied:
Done.

rousskov replied:
I do not think this is done. I see many `merged run`references (and only one `staged run` reference): see `git grep merged.run`

eduard-bagdasaryan replied:
Fixed at 3aa5ac. I renamed into 'staged_run' (not 'staged-run'), to comply with other parameters.

rousskov commented on config-example.json:
```diff
@@ -0,0 +1,28 @@
+    "necessary_approvals": 1,
+    "sufficient_approvals": 2,
+    "voting_delay_min": 48,
+    "voting_delay_max": 240,
```
Please use [timestring](https://github.com/mike182uk/timestring) for parsing time-related configuration values and use human-friendly values in the examples/documentation. README should point to that library instead of describing format details. The configuration values will become strings.

eduard-bagdasaryan replied:
Done.

rousskov commented on MergeContext.js:
```diff
@@ -0,0 +1,506 @@
+const assert = require('assert');
+const endOfLine = require('os').EOL;
```
This does not look right: The newline character in generated git commit messages should not depend on the OS. It should be whatever git or GutHub wants, right? Even if git/GitHub translates `\r\n` into `\n` automatically, then I would still just use `\n` whenever we need a newline character.

eduard-bagdasaryan replied:
Done.

rousskov commented on MergeContext.js:
```diff
@@ -0,0 +1,506 @@
+    _prHeadSha() { return this._pr.head.sha; }
+
+    _prMessage() {
+        return this._pr.title + endOfLine + this._pr.body + endOfLine +
```
PR number goes after PR title, not after the comment body. See numerous examples in Squid master commit log. We are mimicking what GitHub does here.

eduard-bagdasaryan replied:
Done.

rousskov commented on README.md:
```diff
@@ -1,68 +1,302 @@
+## Commit message
+
+The staging commit message and, hence, the target branch commit is
+formed by concatenating the PR title (with an appended PR number), and
```
s/and empty/an empty/

eduard-bagdasaryan replied:
Done.

rousskov commented on README.md:
```diff
@@ -1,68 +1,302 @@
+
+Neither the title nor the description are currently processed to convert
+GitHub markdown to plain text. However, both texts must conform to the
+72 characters/line limit. PRs violating this limit are labeled
```
We check the commit message lines _after_ adding the PR number. This is probably the right thing to do, but it places more stringent requirements on the title. The length of the added ` (#NNN)` title suffix is 7. Let's add something like this after the `However, ...` sentence:

```
The automatically added ` #(NNN)` title suffix further reduces the maximum PR title length
to ~65 characters.
```

and s/violating this limit/violating these limits/.

eduard-bagdasaryan replied:
Done.

rousskov commented on MergeContext.js:
```diff
@@ -0,0 +1,506 @@
+
+    _prMessage() {
+        return this._pr.title + endOfLine + this._pr.body + endOfLine +
+            "(PR #" + this._pr.number + ")";
```
Drop the `PR` part. Again, see actual GitHub commits for examples.

eduard-bagdasaryan replied:
Done.

rousskov commented on PrMerger.js:
```diff
@@ -0,0 +1,92 @@
+            return true; // still in-process
+
+        const prList = await GH.getPRList();
+        prList.sort((pr1, pr2) => { return new Date(pr1.created_at) - new Date(pr2.created_at); });
```
I doubt PR creation dates are guaranteed to be unique. They are also not as easily accessible as PR numbers. PR numbers are probably unique. Let's use PR numbers/IDs instead.

eduard-bagdasaryan replied:
Done.

rousskov commented on Util.js:
```diff
@@ -0,0 +1,83 @@
+
+function MergingTag(prNum) {
+    assert(prNum);
+    return "tags/T-merging-PR" + prNum;
```
Please s/T-merging-PR/M-staged-PR/.

eduard-bagdasaryan replied:
Done.

rousskov reviewed (changes requested):
Reviewed MergingTag(), spotted a few random bugs, and added a few random polishing requests. This is not a comprehensive review, but it should allow you to move forward.

eduard-bagdasaryan committed:
Various fixes after review

* use 'timestring' library for parsing 'delay' parameters
* use 'staging tag' notion instead of 'merge tag' because 'merge tag'
  can be wrongly associated with GitHub-generated 'merge commit'
* fixed 'staging tag' naming
* properly use PR numbers in the staging commit message
* refactored MergeContext._stagingFailed(): do not confuse a reviewer,
  returning a result for a possibly stale tag
* sort PR list by PR numbers (it is possibly better than sorting by
  dates)
* fixed several README wordings

eduard-bagdasaryan committed:
Do not run all 'preconditions' checks as 'postconditions'

Commit message should be checked only before 'staging tag' creation.
Also do not do useless 'stagingFailed()' check at postcondition check
step.
Also reset all labels just before start processing a PR: this is needed
to avoid possible confusion with stale labels.

eduard-bagdasaryan committed:
Updated deprecated dependency: 'github' -> '@octokit/rest'

eduard-bagdasaryan committed:
A couple of fixes

* Do not reset labels in 'dry_run' mode
* Simplified informing about 'dry' modes

eduard-bagdasaryan committed:
Renamed merged_run -> staged_run

eduard-bagdasaryan committed:
Properly assign an author for staged commit

Now, when 'stated commit' is created, the author information is
copied from the GitHub-generated PR 'merge commit'(GitHub fills
these fields automatically with PR creator login and email).

Also fixed a couple of minor problems found during testing.

eduard-bagdasaryan committed:
Set explicitly 'committer' for staged commit

According to the API documentation
https://developer.github.com/v3/git/commits/#create-a-commit,
the 'committer' section is optional and filled with 'author' data
(if the author section is specified). Since we now provide the author
(effectively the PR author), we need to specify manually committer
fields (which is the bot itself).

eduard-bagdasaryan committed:
Use bot 'primary email' for 'committer' in staged commit

... instead of 'noreply' email (21e8280).

eduard-bagdasaryan committed:
Trim status contexts before comparing

Status context is a string filled by a user in a CI configuration.
Accidentally, such strings may contain extra spaces (added by a
mistake).  The bot should not stumble on such contexts and
compare them correctly.

rousskov committed:
Documented the upcoming staging_checks configuration parameter

Staging checks may differ from target branch checks when projects
minimize the number of expensive/slow tests or when the tools like
Jenkins GitHub plugins report PR tests and branch tests differently.

We have considered using GitHub staging branch configuration to store
which tests are required, but that means we have to enable GitHub
protection for the staging branch. We are not sure whether enabling
branch protection would prevent the bot from reseting the staging branch
as needed. We may redesign if testing proves these worries unfounded.

eduard-bagdasaryan committed:
Distinguish GtiHub user 'name' and 'login' properties

We should use 'name' property when specifying staged commit
'name' attribute instead of 'login'.

To avoid confusion, renamed configuration 'github_username' into
'github_login'. This parameter is used for authentication.

eduard-bagdasaryan committed:
Distinguish branch names and 'branch paths'

No functionality changes.

eduard-bagdasaryan committed:
Properly remove stale labels

When the bot starts processing a PR, having stale labels,
it should distinguish two different groups of such labels:

* Labels which should be cleared immediately because they are not
  actual anymore (M-waiting-staging-checks and M-passed-staging checks)

* Labels which should be cleared only after preconditions check passing
  (M-failed-staging-checks and M-failed-other)

eduard-bagdasaryan committed:
Treat 'WIP:' PR title prefix as merge pre/post conditions

eduard-bagdasaryan committed:
Implemented functionality for staging_checks parameter

This parameter was added at d996d41.

Also three bugs were fixed:

* A getCommit() caller treated returned value incorrectly because its
  semantics was changed in fc10e9f.
* There is one case when _cleanupMergeFailed() should not delete failed
  tag. Actually, the tag should be deleted only after a PR or base
  branch change.
* _cleanupMergeFailed() improperly called the labels cleanup method,
  passed as a parameter.

eduard-bagdasaryan committed:
Do not add/remove already existent/non-existent labels

rousskov commented:
Closing as stale, eventually replaced by https://github.com/measurement-factory/anubis/pull/25.
